src/stale-detection.js

```diff
@@ -0,0 +1,81 @@
+import { getIncomingHeartbeat, getLastRecordedIncomingMirror, getLastRecordedOutgoingMirror } from './peer-state.js';
+import { getUnixTimestamp, logInfo, sleep } from './utils.js';
+import { isRunningHeartbeatForPeer } from './heartbeats.js';
+import {
+  GRACE_PERIOD_IN_SECONDS_TO_MIRROR_INCOMING_DRIVE,
+  GRACE_PERIOD_IN_SECONDS_TO_MIRROR_OUTGOING_DRIVE,
+  GRACE_PERIOD_IN_SECONDS_TO_RECEIVE_NEW_HEARTBEAT,
+  HEARTBEAT_INTERVAL_IN_SECONDS,
+  STALENESS_CHECK_INTERVAL_IN_SECONDS,
+} from './constants.js';
+
+export async function detectOutgoingStaleness(peer, { mirrorOutgoing, onStaleDriveDetected }) {
+  // We break out of the loop if staleness is detected
+  while (true) {
+    await sleep(STALENESS_CHECK_INTERVAL_IN_SECONDS * 1000);
+    if (!isRunningHeartbeatForPeer(peer)) {
```

Comment:
Why do we continue detecting staleness for peers for whom we stopped running heartbeats?

Reply:
The `completeHyperdriveSetup` function only gets called the first time a peer connects, and we set up these stale detectors in that `completeHyperdriveSetup` function. So these stale detectors are like the change detectors, they're always on (until staleness is detected).